corehq/motech/fhir/views.py

```diff
@@ -208,5 +222,62 @@ def _validate_case_relationships(self):
             raise Http404(_("There are no cases associated with this user"))
 
 
+class SmartTokenView(TokenView):
+    urlname = "smart_token_view"
+
+    def post(self, request, domain, *args, **kwargs):
+        if request.POST.get('grant_type') == "refresh_token":
+            return self._handle_refresh_token(request, *args, **kwargs)
+
+        response = super().post(request, *args, **kwargs)
+
+        if response.status_code != 200:
+            return response
+
+        body = json.loads(response.content)
+        access_token = body.get("access_token")
+        token = get_access_token_model().objects.get(token=access_token)
+
+        if "launch/patient" in token.scope:
+            authorization_code = request.POST.get('code')
+            client_id = request.POST.get('client_id')
+            case_id = cache.get(_get_smart_cache_key(domain, client_id, authorization_code))
```

Comment:
where does this get set in the cache?

Reply:
[In the first commit](https://github.com/dimagi/commcare-hq/pull/29504/commits/b9a38a5bf75182d8aa60694947f0d066269e75a3#diff-62aec94693d6a2a0bc3596babebc35158c707467255d734c843df469c71809ecR200).

Reply:
Yeah, it happens in the Auth view. With the way I rebased these commits it may have become a bit confusing, sorry!

Those two views get hit in succession, but from completely different places. The Auth view gets submitted a form _from HQ_, which then sends back the authorization code. In that form we allow the user to select a case, and it is the only place this information is available to us. The client then hits this view, using the authorization code to request the access token for the user. Once that access token is created here, we link it with the case that was selected in the Auth view.

Reply:
makes sense, thanks both!